### PR TITLE
os/binfmt: Refractoring binary loader

### PR DESCRIPTION
**Pull request by kishore-sn**
- Remove unused and unnecessary variables
- Make changes to mpu and section allocations
- Remove addrenv since we dont use it
- Some more cleanup

Signed-off-by: Kishore S N <kishore.sn@samsung.com>

### CODE REVIEW

File: os/include/tinyara/binfmt/binfmt.h [hunk 1]
```diff
@@ -75,16 +75,16 @@
...
-	ALLOC_DTOR,
+	BIN_CTOR,
+	BIN_DTOR,
 #endif
-#ifdef CONFIG_OPTIMIZE_APP_RELOAD_TIME
-	ALLOC_RO,
-	ALLOC_DATA,
-#endif
-	ALLOC_MAX
+	BIN_RO,
```
Comment (jeongchanKim) [inline os/include/tinyara/binfmt/binfmt.h:83]: I think `BIN_RO` is used only when `CONFIG_OPTIMIZE_APP_RELOAD_TIME` is enabled.
Could you check this?

File: os/binfmt/binfmt_exit.c [hunk 2]
```diff
@@ -103,12 +99,11 @@ extern volatile sq_queue_t g_delayed_kufree;
 int binfmt_exit(FAR struct binary_s *bin)
 {
 	int ret;
-#ifdef CONFIG_APP_BINARY_SEPARATION
 	irqstate_t flags;
 	FAR void *address;
 	uint32_t uheap_start;
 	uint32_t uheap_end;
-#endif
+
```
Comment (jeongchanKim) [inline os/binfmt/binfmt_exit.c:106]: Please remove unnecessary.

File: os/binfmt/Kconfig [hunk 3]
```diff
@@ -24,6 +24,7 @@ config PATH_INITIAL
 config BINFMT_LOADABLE
 	bool
 	default n
+	depends on APP_BINARY_SEPARATION
```
Comment (jeongchanKim) [inline os/binfmt/Kconfig:27]: `CONFIG_APP_BINARY_SEPARATION` selects the `CONFIG_BINFMT_LOADABLE`.
It looks dangling.
Could you check this?

File: os/binfmt/libelf/libelf_bind.c [hunk 4]
```diff
@@ -441,42 +425,19 @@ int elf_bind(FAR struct elf_loadinfo_s *loadinfo, FAR const struct symtab_s *exp
...
-	up_coherent_dcache(loadinfo->textalloc, loadinfo->textsize);
-	up_coherent_dcache(loadinfo->dataalloc, loadinfo->datasize);
+	up_coherent_dcache(loadinfo->binp->sections[BIN_TEXT], loadinfo->binp->sizes[BIN_TEXT]);
+	up_coherent_dcache(loadinfo->binp->sections[BIN_DATA], loadinfo->binp->sizes[BIN_DATA]);
 
 #endif
 
-#if defined(CONFIG_ARCH_ADDRENV) || defined(CONFIG_SUPPORT_COMMON_BINARY)
+#ifdef CONFIG_SUPPORT_COMMON_BINARY
 ret_err:
```
Comment (jeongchanKim) [inline os/binfmt/libelf/libelf_bind.c:440]: Originally, `strtab` and `symtab` will be freed regardless of `ret_err`.
But with your changes, `strtab` and `symtab` will be freed only when `CONFIG_SUPPORT_COMMON_BINARY` is enabled.
Could you check this?

File: os/binfmt/libelf/libelf_read.c [hunk 5]
```diff
@@ -117,10 +117,7 @@ static inline void elf_dumpreaddata(FAR char *buffer, int buflen)
  *
  * Description:
  *   Read 'readsize' bytes from the object file at 'offset'.  The data is
- *   read into 'buffer.' If 'buffer' is part of the ELF address environment,
- *   then the caller is responsibile for assuring that that address
- *   environment is in place before calling this function (i.e., that
- *   elf_addrenv_select() has been called if CONFIG_ARCH_ADDRENV=y).
+ *   read into 'buffer.'
```
Comment (jeongchanKim) [inline os/binfmt/libelf/libelf_read.c:120]: Nothing important, but how about `'buffer'.`?

File: os/binfmt/binfmt.h [hunk 6]
```diff
@@ -148,6 +151,51 @@ void binfmt_freeargv(FAR struct binary_s *bin);
...
+static inline void binfmt_mpuinit(struct binary_s *binp)
+{
+	uint32_t *regs = NULL;
+	uint8_t nregion = 0;
+
+#ifdef CONFIG_SUPPORT_COMMON_BINARY
+	if (binp->islibrary) {
+		regs = binp->cmn_mpu_regs;
+		nregion = mpu_get_nregion_info(MPU_REGION_COMMON_BIN);
+	} else
```
Comment (sunghan-chang) [inline os/binfmt/binfmt.h:165]: `else` looks strange.
I think you intended two cases as below:
1. remove `#else`
```
#ifdef config_common_bin
   if (islibrary) {
       do common bin
   } else
#endif 
   {
       do app bin
   }
```
2. remove `else`
```
#ifdef config_common_bin
   if (islibrary) {
       do common bin
   } 
#else 
   {
       do app bin
   }
#endif
```

I think 1st is correct. Could you check this?

File: os/binfmt/elf.c [hunk 7]
```diff
@@ -121,10 +121,10 @@ static void elf_dumploadinfo(FAR struct elf_loadinfo_s *loadinfo)
...
 
 	binfo("LOAD_INFO:\n");
-	binfo("  textalloc:    %08lx\n", (long)loadinfo->textalloc);
-	binfo("  dataalloc:    %08lx\n", (long)loadinfo->dataalloc);
-	binfo("  textsize:     %ld\n", (long)loadinfo->textsize);
-	binfo("  datasize:     %ld\n", (long)loadinfo->datasize);
+	binfo("  textalloc:    %08lx\n", (long)binp->sections[BIN_TEXT]);
+	binfo("  dataalloc:    %08lx\n", (long)binp->sections[BIN_DATA]);
+	binfo("  textsize:     %ld\n", (long)binp->sizes[BIN_TEXT]);
+	binfo("  datasize:     %ld\n", (long)binp->sizes[BIN_DATA]);
```
Comment (sunghan-chang) [inline os/binfmt/elf.c:127]: Where is `binp`? Is it global variable?

**Review by sunghan-chang — COMMENTED** [review]
If we apply this PR, there is no difference between common and app for code location and sequence? Could you confirm this?

File: os/binfmt/binfmt_execmodule.c [hunk 8]
```diff
@@ -150,18 +150,11 @@ static void exec_ctors(FAR void *arg)
 int exec_module(FAR struct binary_s *binp)
 {
 	FAR struct task_tcb_s *tcb;
```
Comment (sunghan-chang) [inline os/binfmt/binfmt_execmodule.c:152]: How about renaming this to distinguish rtcb?
loading_stcb? (s : start)
main_tcb? mtcb?

File: os/binfmt/binfmt_execmodule.c [hunk 9]
```diff
@@ -333,12 +272,14 @@ int exec_module(FAR struct binary_s *binp)
 
 	pid = tcb->cmn.pid;
 
-#ifdef CONFIG_APP_BINARY_SEPARATION
 	rtcb->uheap = 0;
+#ifdef CONFIG_ARM_MPU						/* MPU register values for loadable apps only */
+	memset(rtcb->mpu_regs, 0, sizeof(rtcb->mpu_regs));
```
Comment (sunghan-chang) [inline os/binfmt/binfmt_execmodule.c:277]: Could you leave why those two lines are necessary?

File: os/binfmt/binfmt_exit.c (hunk 2 above)
Comment (sunghan-chang) [reply os/binfmt/binfmt_exit.c:106]: Let's remove a tab.

File: os/binfmt/binfmt_exit.c [hunk 10]
```diff
@@ -136,7 +130,7 @@ int binfmt_exit(FAR struct binary_s *bin)
 		address = (FAR void *)sq_next((FAR sq_entry_t *)address);
 	}
 	mm_disable_app_heap_list(bin->uheap);
-#endif
+
```
Comment (sunghan-chang) [inline os/binfmt/binfmt_exit.c:133]: same as previous

File: os/binfmt/binfmt_execmodule.c (hunk 8 above)
Comment (kishore-sn) [reply os/binfmt/binfmt_execmodule.c:152]: Ok. I will rename it as newtcb.

File: os/binfmt/binfmt_execmodule.c (hunk 9 above)
Comment (kishore-sn) [reply os/binfmt/binfmt_execmodule.c:277]: Done

File: os/binfmt/binfmt_exit.c (hunk 2 above)
Comment (kishore-sn) [reply os/binfmt/binfmt_exit.c:106]: Done

File: os/binfmt/binfmt_exit.c (hunk 10 above)
Comment (kishore-sn) [reply os/binfmt/binfmt_exit.c:133]: Done.

File: os/binfmt/binfmt_execmodule.c [hunk 11]
```diff
@@ -325,33 +264,36 @@ int exec_module(FAR struct binary_s *binp)
...
 
-	pid = tcb->cmn.pid;
+	pid = newtcb->cmn.pid;
 
-#ifdef CONFIG_APP_BINARY_SEPARATION
 	rtcb->uheap = 0;
+#ifdef CONFIG_ARM_MPU
+	/* Reset the mpu_regs in parent task to zero */
+	memset(rtcb->mpu_regs, 0, sizeof(rtcb->mpu_regs));
+#endif
```
Comment (sunghan-chang) [inline os/binfmt/binfmt_execmodule.c:279]: Could you review to set new_tcb here instead of getting them from rtcb and clearing?

Comment (sunghan-chang) [reply os/binfmt/binfmt_execmodule.c:279]: If you want to keep original, then let's leave why parent kernel thread needs to be initialized.

**Comment by sunghan-chang** [comment]
Could you fix typo, Refractoring in commit title?

File: os/binfmt/elf.c [hunk 12]
```diff
@@ -121,10 +121,10 @@ static void elf_dumploadinfo(FAR struct elf_loadinfo_s *loadinfo)
 	int i;
 
 	binfo("LOAD_INFO:\n");
-	binfo("  textalloc:    %08lx\n", (long)loadinfo->textalloc);
-	binfo("  dataalloc:    %08lx\n", (long)loadinfo->dataalloc);
-	binfo("  textsize:     %ld\n", (long)loadinfo->textsize);
-	binfo("  datasize:     %ld\n", (long)loadinfo->datasize);
+	binfo("  textalloc:    %08lx\n", (long)loadinfo->binp->sections[BIN_TEXT]);
+	binfo("  dataalloc:    %08lx\n", (long)loadinfo->binp->sections[BIN_DATA]);
```
Comment (jeongchanKim) [inline os/binfmt/elf.c:125]: How about printing ro also?

File: os/binfmt/libelf/libelf_sections.c [hunk 13]
```diff
@@ -188,9 +188,9 @@ static inline int elf_sectname(FAR struct elf_loadinfo_s *loadinfo, FAR const El
 #ifdef CONFIG_BINFMT_SECTION_UNIFIED_MEMORY
 void *elf_find_start_section_addr(struct binary_s *binp)
 {
-	int text_addr = (int)binp->alloc[ALLOC_TEXT];
-	int ro_addr = (int)binp->alloc[ALLOC_DATA];
-	int data_addr = (int)binp->alloc[ALLOC_DATA];
+	int text_addr = (int)binp->sections[BIN_TEXT];
+	int ro_addr = (int)binp->sections[BIN_DATA];
```
Comment (jeongchanKim) [inline os/binfmt/libelf/libelf_sections.c:192]: Is it correct?
I think we should use `BIN_RO` instead of `BIN_DATA`.

**Comment by sunghan-chang** [comment]
@kishore-sn Could you check below?
```
[sunghan-chang] commented [4 days ago]
Could you fix typo, Refractoring in commit title?
```

**Comment by sunghan-chang** [comment]
@kishore-sn Could you check build results with config combinations? CONFIG_OPTIMIZE_APP_RELOAD_TIME enable / disable, CONFIG_SUPPORT_COMMON_BINARY enable / disable

**Comment by kishore-sn** [comment]
> @kishore-sn Could you check below?
> 
> ```
> [sunghan-chang] commented [4 days ago]
> Could you fix typo, Refractoring in commit title?
> ```

Done

**Comment by kishore-sn** [comment]
> @kishore-sn Could you check build results with config combinations? CONFIG_OPTIMIZE_APP_RELOAD_TIME enable / disable, CONFIG_SUPPORT_COMMON_BINARY enable / disable

Yes. I have already checked these. They are working well.

**Review by sunghan-chang — APPROVED** [review]

**Pull request merged**